### PR TITLE
rust agents: expose prometheus metrics

### LINKED ISSUE #717: metrics: Implement Metrics Scaffolding
here's the implementation sketch

- [ ] Add a `metrics` mod to `optics-base`
- [ ] metrics mod contains once_cell for a `CoreMetrics` struct
- [ ] `CoreMetrics` contains a selection of metrics and exposes a `gather`
    - tx counts
    - wallet balance
    - blockheight
    - wallet address 
- [ ] Define an `AgentMetrics` in each agent with `once_cell` in a `metrics.rs`
    - macroify (see `decl_settings` macro)
- [ ] `AgentsMetrics` exposes agent-specific metrics and exposes a `gather`
    - e.g. updater attestations signed and total
    - e.g. current `next_to_inspect` for processor

_Originally posted by @prestwich in https://github.com/celo-org/optics-monorepo/discussions/645#discussioncomment-1241232_

### PR DESCRIPTION
**Pull request by emberian**
Progress on #657
Closes #717

### CODE REVIEW

**Comment by emberian** [comment]
@prestwich this isn't quite what you spec'd out, when it comes to the agent details, but i think it's lower boilerplate. how do you think this looks?

File: rust/agents/updater/src/main.rs [hunk 1]
```diff
@@ -22,6 +23,17 @@ async fn _main() -> Result<()> {
...
 
     let agent = Updater::from_settings(settings).await?;
+
+    let metrics = agent.metrics();
+    let metrics_endpoint =
+        warp::path!("metrics").map(move || metrics.gather().expect("failed to encode metrics"));
+
+    tokio::spawn(async move {
+        warp::serve(metrics_endpoint)
+            .run(([0, 0, 0, 0], 8080))
```
Comment (yourbuddyconner) [inline rust/agents/updater/src/main.rs:33]: Please use port 9090 by default here, allowing it to be set using the config pipeline would be nice too.

**Comment by emberian** [comment]
afaict the only major problem with this right now is running multiple agents in the same binary is unreasonable (can't merge the registries/HTTP servers).

**Comment by yourbuddyconner** [comment]
> afaict the only major problem with this right now is running multiple agents in the same binary is unreasonable (can't merge the registries/HTTP servers).

I dont think we anticipate ever running multiple instances of an agent from a single binary, so this is an acceptable constraint.

**Comment by prestwich** [comment]
I am generally liking what I see. Can we run from this branch and confirm that the endpoint works?

File: rust/optics-base/src/settings/mod.rs [hunk 2]
```diff
@@ -196,7 +196,9 @@ impl Settings {
             db,
             metrics: Arc::new(crate::metrics::CoreMetrics::new(
                 name,
-                self.metrics,
+                self.metrics
+                    .as_ref()
+                    .map(|v| v.parse::<u16>().expect("metrics port must be u16")),
```
Comment (prestwich) [inline rust/optics-base/src/settings/mod.rs:201]: can we do a reasonable default and `unwrap_or` ?  

cc @yourbuddyconner what would a reasonable default metrics port be?

Comment (yourbuddyconner) [reply rust/optics-base/src/settings/mod.rs:201]: Port 9090 is default prometheus port

Comment (emberian) [reply rust/optics-base/src/settings/mod.rs:201]: i would rather have a way to disable the metrics server.

the reasonable default is in the default config, is that a bad place for it?

Comment (prestwich) [reply rust/optics-base/src/settings/mod.rs:201]: I am hard pressed to think of a situation where we would want to disable metrics. Do you have a use case in mind?

Comment (emberian) [reply rust/optics-base/src/settings/mod.rs:201]: i probably wouldn't use metrics on my own xapp agent and until this PR i never got the popup from macOS to punch a hole through the firewall for a listening port. not particularly beyond operator paranoia.

**Comment by yourbuddyconner** [comment]
Screen Shot 2021-09-03 at 12 33 26 PM

Verified that this endpoint loads in a dev environment, haven't tested whether metrics actually increment. We can probably merge this scaffolding and then verify individual metrics in subsequent PRs.

**Comment by yourbuddyconner** [comment]
This PR closes #717

File: rust/optics-base/src/settings/mod.rs (hunk 2 above)
Comment (yourbuddyconner) [reply rust/optics-base/src/settings/mod.rs:201]: Should definitely be switchable. Default should be off, and we should include a default config for port 9090 in the config generator as we will never run agents without metrics, and neither will anyone we know.

**Comment by yourbuddyconner** [comment]
Also verified that prometheus scrapes this endpoint alright. Think this scaffolding PR is good to go. 

Screen Shot 2021-09-03 at 12 55 38 PM

**Review by yourbuddyconner — APPROVED** [review]

**Pull request merged**